KAFKA-6367: StateRestoreListener use actual last restored offset for restored batch

bbejeck opened the pull request:
Use the last actual restored offset for the `StoreRestoreListener.onBatchRestored` method.  Probably should update the docs to inform users this could include gaps in sequence due to commit markers.

Updated existing tests

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

bbejeck committed:
KAFKA-6367: StateRestoreListener use actual last restored offset for restored batch

bbejeck commented:
@dguy @guozhangwang @mjsax

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -273,8 +273,9 @@ private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
+        long offset = 0;
         for (final ConsumerRecord<byte[], byte[]> record : records) {
-            final long offset = record.offset();
+            offset = record.offset();
```
I think this is only correct, if `record` is included in the restore. However, if `hasCompleted` returns `true` the record will not be restored and we need to use the offset of the record before?

bbejeck replied:
ack

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/260/

bbejeck committed:
KAFKA-6367: update to correctly report last restored offset if completes before all records in batch are exhausted.

bbejeck commented:
updated this

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/293/

bbejeck commented:
retest this please

mjsax commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java:
```diff
@@ -273,12 +273,14 @@ private long processNext(final List<ConsumerRecord<byte[], byte[]>> records,
         long nextPosition = -1;
         int numberRecords = records.size();
         int numberRestored = 0;
+        long lastRestoredOffset = 0;
```
Nit: init with `-1` (does not really change much, but indicates "unset" better than zero).

bbejeck replied:
ack

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java:
```diff
@@ -234,15 +234,28 @@ public void shouldRestoreAndNotifyMultipleStores() throws Exception {
 
         assertAllCallbackStatesExecuted(callback, "storeName1");
-        assertCorrectOffsetsReportedByListener(callback, 0L, 10L, 10L);
+        assertCorrectOffsetsReportedByListener(callback, 0L, 9L, 10L);
```
should `endOffset` be `9L`, too?

bbejeck replied:
actually, this parameters and variables within the verification need to change.  The last parameter is for the number of records restored.  The last offset restored prior to this PR before was incorrect, it was the offset of the next record, not the offset of the actual last record restored.

mjsax replied:
Ok. That makes sense.

Nevertheless, the test passed comparing `restoreListener.restoreEndOffset, equalTo(endOffset)` -- thus, the `MockStateRestoreListener#restoreEndOffset` is set to `10L` -- this sound wrong.

What I try to say is, that this bug does affect the batch-end-offset and the overall-end-offset. Thus, we should update the Jira description and also add an fix to this PR for the overall-end-offset.

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java:
```diff
@@ -234,15 +234,28 @@ public void shouldRestoreAndNotifyMultipleStores() throws Exception {
     }
 
+    @Test
+    public void shouldOnlyReportTheLastRestoredOffset() {
```
Great test!

bbejeck committed:
KAFKA-6367: updates per comments

bbejeck commented:
updated this

mjsax commented on streams/src/test/java/org/apache/kafka/streams/processor/internals/StoreChangelogReaderTest.java:
```diff
@@ -253,11 +266,11 @@ private void assertAllCallbackStatesExecuted(final MockStateRestoreListener rest
     private void assertCorrectOffsetsReportedByListener(final MockStateRestoreListener restoreListener,
                                                         final long startOffset,
-                                                        final long batchOffset, final long endOffset) {
+                                                        final long batchOffset, final long totalRestored) {
```
nit: new line

bbejeck committed:
KAFKA_6367:updates per comments

bbejeck commented:
>What I try to say is, that this bug does affect the batch-end-offset and the overall-end-offset. Thus, we should update the Jira description and also add a fix to this PR for the overall-end-offset.

@mjsax I believe was just the test naming conventions that were incorrect, I've updated the test to correct this.  The overall end-offset is recorded in the beginning in the first call to the listener.  I've also updated the Jira to a more accurate title/description

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/343/

guozhangwang commented:
retest this please

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/353/

bbejeck committed:
KAFKA-6367: updated java docs for correct semantics

mjsax reviewed (approved):
LGTM.

mjsax commented:
Merged to `trunk` and cherry-picked to `1.1` and `1.0` branches.
